doc/developer/design/20240625_source_versioning__table_from_sources.md

```diff
@@ -0,0 +1,537 @@
+
+Note that we do not expose the types of any of the upstream columns, since each source has
+different constructs for how types are represented, and since some source-types can't
+represent all upstream types (e.g. MySQL) without knowing which of them will be represented
+as TEXT using a TEXT COLUMNS option. Therefore the list of `columns` does not imply that each
+of those columns is one that Materialize can actually ingest.
+
+A new `REFRESH SOURCE <source_id or name>` statement will be introduced that can trigger
+a refresh of the rows in `mz_catalog.mz_available_source_references` that pertain to the
+requested source. Otherwise, this catalog table will only be populated at source creation
+time and upon a defined refresh interval in the source operator (perhaps every 60 seconds).
+
+### System Catalog & Introspection Updates
+
+We currently surface details about both Sources and Tables through various system catalog
+tables in both the `mz_catalog` and `mz_internal` schemas. These expose information
+about the objects themselves as well as statistics and statuses emitted from the active dataflows for each source.
+
+We would make the following updates to provide users and Materialize employees the same
+ability to introspect source-fed tables that they currently have for subsources.
+
+#### [mz_catalog.mz_tables](https://materialize.com/docs/sql/system-catalog/mz_catalog/#mz_tables)
+
+`CREATE TABLE .. FROM SOURCE` statements would be added to this catalog table.
+
+Two new columns would be added:
+
+`data_source_type`: distinguishes these tables
+from 'normal' tables that are written to directly by users. The value of this will
+be one of `'direct'` or `'source'`.
+
+`data_source_id`: a nullable column that contains the `id` of the data source,
+if there is one. In the case of `CREATE TABLE .. FROM SOURCE` statements this would
+contain the `id` of the primary source.
+
+#### [mz_catalog.mz_sources](https://materialize.com/docs/sql/system-catalog/mz_catalog/#mz_sources)
+
+`CREATE TABLE .. FROM SOURCE` statements will be added to this catalog table, despite
+them being 'tables', since this catalog table contains columns that are unique to
+source-related objects (e.g. `connection_id`).
+
+We will add the variant `'table'` to the `type` column which is currently a `text` column
+containing one of:
+`'kafka', 'mysql', 'postgres', 'load-generator', 'progress', or 'subsource'.`
+
+Once all customer's sources have been migrated to the new model, the `'subsource'` variant will be removed from the `type` column.
+
+Open Question: Will it be confusing to users that these source-fed tables will appear
+in both the `mz_sources` and `mz_tables` catalog tables?
+
+#### [mz_internal.mz_postgres_source_tables](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_postgres_source_tables) and [mz_internal.mz_mysql_source_tables](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_mysql_source_tables)
```

Comment:
Do we not need a table for Kafka, too, which maps the table to the external topic?

Reply:
I understand that the mapping here is 1:1, but it feels like we should expose it in the same way as for PostgreSQL and MySQL.